src/netdef.c

```diff
@@ -0,0 +1,307 @@
+
+    FREE_AND_NULLIFY(netdef->peer);
+
+    netdef->vlan_id = G_MAXUINT; /* 0 is a valid ID */
+    netdef->vlan_link = NULL;
+
+    FREE_AND_NULLIFY(netdef->set_mac);
+    netdef->mtubytes = 0;
+    netdef->ipv6_mtubytes = 0;
+
+    FREE_AND_NULLIFY(netdef->set_name);
+    FREE_AND_NULLIFY(netdef->match.driver);
+    FREE_AND_NULLIFY(netdef->match.mac);
+    FREE_AND_NULLIFY(netdef->match.original_name);
+    netdef->has_match = FALSE;
+    netdef->wake_on_lan = FALSE;
+    netdef->wowlan = 0;
+    netdef->emit_lldp = FALSE;
+
+    if (netdef->access_points) {
+        g_hash_table_foreach_remove(netdef->access_points, free_access_point, &backend);
+        g_hash_table_destroy(netdef->access_points);
+        netdef->access_points = NULL;
+    }
+
+    FREE_AND_NULLIFY(netdef->bond_params.mode);
+    FREE_AND_NULLIFY(netdef->bond_params.lacp_rate);
+    FREE_AND_NULLIFY(netdef->bond_params.monitor_interval);
+    FREE_AND_NULLIFY(netdef->bond_params.transmit_hash_policy);
+    FREE_AND_NULLIFY(netdef->bond_params.selection_logic);
+    FREE_AND_NULLIFY(netdef->bond_params.arp_interval);
+    free_garray_with_destructor(&netdef->bond_params.arp_ip_targets, g_free);
+    FREE_AND_NULLIFY(netdef->bond_params.arp_validate);
+    FREE_AND_NULLIFY(netdef->bond_params.arp_all_targets);
+    FREE_AND_NULLIFY(netdef->bond_params.up_delay);
+    FREE_AND_NULLIFY(netdef->bond_params.down_delay);
+    FREE_AND_NULLIFY(netdef->bond_params.fail_over_mac_policy);
+    FREE_AND_NULLIFY(netdef->bond_params.primary_reselect_policy);
+    FREE_AND_NULLIFY(netdef->bond_params.learn_interval);
+    FREE_AND_NULLIFY(netdef->bond_params.primary_slave);
+    memset(&netdef->bond_params, 0, sizeof(netdef->bond_params));
+
+    FREE_AND_NULLIFY(netdef->modem_params.apn);
+    netdef->modem_params.auto_config = FALSE;
+    FREE_AND_NULLIFY(netdef->modem_params.device_id);
+    FREE_AND_NULLIFY(netdef->modem_params.network_id);
+    FREE_AND_NULLIFY(netdef->modem_params.number);
+    FREE_AND_NULLIFY(netdef->modem_params.password);
+    FREE_AND_NULLIFY(netdef->modem_params.pin);
+    FREE_AND_NULLIFY(netdef->modem_params.sim_id);
+    FREE_AND_NULLIFY(netdef->modem_params.sim_operator_id);
+    FREE_AND_NULLIFY(netdef->modem_params.username);
+
+    FREE_AND_NULLIFY(netdef->bridge_params.ageing_time);
+    FREE_AND_NULLIFY(netdef->bridge_params.forward_delay);
+    FREE_AND_NULLIFY(netdef->bridge_params.hello_time);
+    FREE_AND_NULLIFY(netdef->bridge_params.max_age);
+    memset(&netdef->bridge_params, 0, sizeof(netdef->bridge_params));
+    netdef->custom_bridging = FALSE;
+
+    netdef->tunnel.mode = NETPLAN_TUNNEL_MODE_UNKNOWN;
+    FREE_AND_NULLIFY(netdef->tunnel.local_ip);
+    FREE_AND_NULLIFY(netdef->tunnel.remote_ip);
+    FREE_AND_NULLIFY(netdef->tunnel.input_key);
+    FREE_AND_NULLIFY(netdef->tunnel.output_key);
+    FREE_AND_NULLIFY(netdef->tunnel.private_key);
+    netdef->tunnel.fwmark = 0;
+    netdef->tunnel.port = 0;
```

Comment:
Same here. This is fine, but we could be using `memset()` to keep it in line with the cleanup of the other structs. Then move the `netdef->tunnel.mode = NETPLAN_TUNNEL_MODE_UNKNOWN` after `memset()`.

Reply:
I'm still undecided between `memset()`-ing the substructs vs initializing every component to 0, hence the lack of coherence.

I'll go with `memset()` for now to get this out the doors, but in the future I think I'd rather name the structs and use dedicated `reset_` functions with explicit defaults for each field. It wouldn't add that much more noise anyway :grin:

Reply:
Yeah.. I was leaning towards explicitly initializing every component to 0 at first as well. But OTOH I also liked the forward-compatibility of `memset()` to initialize everything, including future fields, to 0, similar to how `g_new0()` creates this data structure in the first place.
I'd be fine with both, I guess :-)